streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java

```diff
@@ -395,29 +401,42 @@
             JoinWindows windows,
             Serde<K> keySerde,
             Serde<V> thisValueSerde,
-            Serde<V1> otherValueSerde);
+            Serde<V1> otherValueSerde,
+            final String thisStoreName,
+            final String otherStoreName);
 
     /**
      * Combine element values of this stream with another {@link KStream}'s elements of the same key using windowed Inner Join
      * with default serializers and deserializers. If a record key is null it will not included in the resulting {@link KStream}
+     * Both of the joining {@link KStream}s will be materialized in local state stores with the given store names.
+     * Also a changelog topic named "${applicationId}-${storeName}-changelog" will be automatically created
+     * in Kafka for each store for failure recovery, where "applicationID" is user-specified in the
+     * {@link org.apache.kafka.streams.StreamsConfig}.
      *
      * @param otherStream   the instance of {@link KStream} joined with this stream
      * @param joiner        the instance of {@link ValueJoiner}
      * @param windows       the specification of the {@link JoinWindows}
      * @param <V1>          the value type of the other stream
      * @param <R>           the value type of the new stream
-     *
+     * @param thisStoreName name of materialized store for this stream
+     * @param otherStoreName name of materialized store for other stream
      * @return a {@link KStream} that contains join-records for each key and values computed by the given {@link ValueJoiner},
      *         one for each matched record-pair with the same key and within the joining window intervals
      */
     <V1, R> KStream<K, R> join(
             KStream<K, V1> otherStream,
             ValueJoiner<V, V1, R> joiner,
-            JoinWindows windows);
+            JoinWindows windows,
+            final String thisStoreName,
```

Comment:
I hate to regret on decisions we made before.. but when looking at worst-case 8 parameters of the join APIs I still feel not very comfortable, and hence would like to bring up this discussion again (I'm sorry @enothereska ):

I still feel in most cases users would not care about querying stores for materialized streams involved in joins, and perhaps the only reason they want to do it is for debugging, and in this case `stream.print` should be sufficient with timestamps as the querying result is merely just printing the stream contents. And if that is true, not exposing these two state stores may be OK, and we do not need to worry about "we are exposing only a subset of all state stores" -- in fact moving forward we may not materialize them persistently in a key-value store -- since we can always say that "only KTables are possibly materialized and exposed for querying".

What do you think?